### PR TITLE
Covernet baseline

### PR DESCRIPTION
**Pull request by freddyaboulton**
Adds an implementation of the CoverNet model with constant lattice. 

I pulled the backbones out of the mtp module and added them to their own file so they could be shared between MTP and CoverNet.

Added unit tests and a script to test if the covernet implementation can overfit on a constant lattice.

### CODE REVIEW

File: python-sdk/nuscenes/predict/models/covernet.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+        logits = torch.cat([backbone_features, agent_state_vector], dim=1)
+
+        for linear in self.head:
+            logits = linear(logits)
+
+        return logits
+
+
+def l1_distance(lattice: torch.Tensor, ground_truth: torch.Tensor) -> int:
```
Comment (elenacorinagrigore) [inline python-sdk/nuscenes/predict/models/covernet.py]: Looks like this function is returning a tensor, not an int?

File: python-sdk/nuscenes/predict/tests/run_covernet.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    model = model.to(device)
+
+    lattice = torch.zeros(3, 6, 2)
+    lattice[0] = generate_trajectory(math.pi / 2)
+    lattice[1] = generate_trajectory(math.pi / 4)
+    lattice[2] = generate_trajectory(3 * math.pi / 4)
+
+    loss_function = ConstantLatticeLoss(lattice)
+
+    current_loss = 10000
```
Comment (elenacorinagrigore) [inline python-sdk/nuscenes/predict/tests/run_covernet.py]: This statement seems to have no effect (see line 88)?

File: python-sdk/nuscenes/predict/models/backbone.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        :param version: mobilenet version to use.
+        """
+        super().__init__()
+
+        if version != 'mobilenet_v2':
+            raise NotImplementedError(f'Only mobilenet_v2 has been implemented. Received {version}.')
+
+        self.backbone = trim_network_at_index(mobilenet_v2(), -1)
+
+    def forward(self, input_tensor):
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/backbone.py]: Done!

File: python-sdk/nuscenes/predict/models/covernet.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+        super().__init__()
+
+        if not n_hidden_layers:
+            n_hidden_layers = [4096]
+
+        self.backbone = backbone
+
+        backbone_feature_dim = calculate_backbone_feature_dim(backbone, input_shape)
+        n_hidden_layers = [backbone_feature_dim + 3] + n_hidden_layers + [num_modes]
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/covernet.py]: Done!

File: python-sdk/nuscenes/predict/models/covernet.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        return logits
+
+
+def l1_distance(lattice: torch.Tensor, ground_truth: torch.Tensor) -> int:
+    """
+    Computes the index of the closest trajectory in the lattice as measured by l1 distance
+    :param lattice: Lattice of pre-generated trajectories. Shape [num_modes, n_timesteps, state_dim]
+    :param ground_truth: Ground truth trajectory of agent. Shape [1, n_timesteps, state_dim].
+    :return: Index of most closest mode in the lattice.
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/covernet.py]: Went back to first grade English for a second.

File: python-sdk/nuscenes/predict/models/covernet.py (hunk 1 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/covernet.py]: Fixed

File: python-sdk/nuscenes/predict/tests/run_covernet.py (hunk 2 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/tests/run_covernet.py]: Got rid of it!

**Pull request merged**